File: 138.Copy List with Random Pointer/CopyListwithRandomPointer .cpp
/**
 * Definition for singly-linked list with a random pointer.
 * struct RandomListNode {
 *     int label;
 *     RandomListNode *next, *random;
 *     RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
 * };
 */
class Solution {
public:

    int count(RandomListNode *head, RandomListNode *l){
        //count from head to l;
        int i=0;
        if(l!=NULL){
            for(;head!=l;head=head->next){
                i++;
            }
        }
        if(head!=l)
            return -1;
        else
            return i;
    }
    
    RandomListNode *copyRandomList(RandomListNode *head) {
        RandomListNode * myhead=NULL;
        RandomListNode * myl=myhead;
        RandomListNode * l=head;
        //next field's copy.
        l=head;
        while(l!=NULL){
            RandomListNode *p=new RandomListNode(l->label);
            if(myhead==NULL)
            {    
                myhead=p;
                myl=myhead;
            }
            else
            {
                myl->next=p;
                myl=myl->next;
            }
            l=l->next;
        }
        l=head;
        myl=myhead;
        
        while(l!=NULL){
            //copy random field.
            int c=count(head,l->random);
            if(c!=-1){
                RandomListNode *next1=myhead;
                for(int n=0;n!=c;n++){
                    next1=next1->next;
                }
                myl->random=next1;
            }
            l=l->next;
            myl=myl->next;
        }
        
        return myhead;
    }
};

File: 138.Copy List with Random Pointer/CopyListwithRandomPointer.java
/**
 * Definition for singly-linked list with a random pointer.
 * class RandomListNode {
 *     int label;
 *     RandomListNode next, random;
 *     RandomListNode(int x) { this.label = x; }
 * };
 */
public class Solution {
    public RandomListNode copyRandomList(RandomListNode head) {
        Map <RandomListNode,RandomListNode> hash = new HashMap<RandomListNode, RandomListNode>();
        RandomListNode dummy = new RandomListNode(-1);
        RandomListNode last = dummy;
        while(head!= null)
        {
            RandomListNode cur = hash.get(head);
            if(cur == null)
            {
                cur = new RandomListNode(head.label);
            }
            last.next = cur;
            
            if(head.random != null)
            {
                RandomListNode ran = hash.get(head.random);
                if(ran == null)
                {
                    ran = new RandomListNode(head.random.label);
                }
                cur.random = ran;
            }
            last = last.next;
            head = head.next;
        }
        return dummy.next;
    }
}